rush-plugins/rush-http-build-cache-plugin/src/HttpBuildCacheProvider.ts

```diff
@@ -0,0 +1,384 @@
+      body: body,
+      redirect: 'follow'
+    });
+
+    if (!response.ok) {
+      if (typeof credentials !== 'string' && safeCredentialOptions === CredentialsOptions.Optional) {
+        // We tried fetching the resource without credentials and that did not work out
+        // Try again but require credentials this time
+        // This will trigger the provider to request credentials
+        return await this._http({ ...options, credentialOptions: CredentialsOptions.Required });
+      }
+
+      this._reportFailure(terminal, method, response, false, warningText);
+      return false;
+    }
+
+    const result: Buffer | boolean = readBody ? Buffer.from(await response.arrayBuffer()) : true;
+
+    terminal.writeDebugLine(
+      `[http-build-cache] actual response: ${response.status} ${url} ${
+        result === true ? 'true' : result.length
+      } bytes`
+    );
+
+    return result;
+  }
+
+  private async _tryGetCredentials(options: CredentialsOptions.Required): Promise<string>;
+  private async _tryGetCredentials(options: CredentialsOptions.Optional): Promise<string | undefined>;
+  private async _tryGetCredentials(options: CredentialsOptions.Omit): Promise<undefined>;
+  private async _tryGetCredentials(options: CredentialsOptions): Promise<string | undefined>;
+  private async _tryGetCredentials(options: CredentialsOptions): Promise<string | undefined> {
+    if (options === CredentialsOptions.Omit) {
+      return;
+    }
+
+    let credentials: string | undefined = this._environmentCredential;
+
+    if (typeof credentials !== 'string') {
+      credentials = await this._tryGetCredentialsFromCache();
+    }
+
+    if (typeof credentials !== 'string' && options === CredentialsOptions.Required) {
+      throw new Error(
+        [
+          `Credentials for ${this._url.href} have not been provided.`,
+          `In CI, verify that RUSH_BUILD_CACHE_CREDENTIAL contains a valid Authorization header value.`,
+          ``,
+          `For local developers, run:`,
+          ``,
+          `    rush update-cloud-credentials --interactive`,
+          ``
+        ].join('\n')
+      );
+    }
+
+    return credentials;
+  }
+
+  private async _tryGetCredentialsFromCache(): Promise<string | undefined> {
+    let cacheEntry: ICredentialCacheEntry | undefined;
+
+    await CredentialCache.usingAsync(
+      {
+        supportEditing: false
+      },
+      (credentialsCache: CredentialCache) => {
+        cacheEntry = credentialsCache.tryGetCacheEntry(this._credentialCacheId);
+      }
+    );
+
+    if (cacheEntry) {
+      const expirationTime: number | undefined = cacheEntry.expires?.getTime();
+      if (!expirationTime || expirationTime >= Date.now()) {
+        return cacheEntry.credential;
+      }
+    }
+  }
+
+  private _getFailureType(requestMethod: string, response: Response, isRedirect: boolean): FailureType {
+    if (response.ok) {
+      return FailureType.None;
+    }
+
+    switch (response.status) {
+      case 503: {
```

Comment:
Seems like this should be for any 5xx error.

Reply:
So I ended up adding a comment instead of changing this -- our testing is all on S3, where there's a couple types of failures we are catching (Service Unavailable, Rate Limit Throttle), which we know are transient.

But even according to HTTP Status spec, 500 = Unknown Error, 501 = Not Implemented, 502 = Bad Gateway, 503 = Service Unavailable. So I think our assumption is that a generic 5xx error is actually most likely a configuration problem in the plugin and _should_ be reported (failed build), whereas a 503 is likely temporary and can just be printed.